### PR TITLE
bpo-32912: Upgrade warning for invalid escape sequences from silent to non-silent

### PR DESCRIPTION
**Pull request by Vgr255**
What the title says. Properly tested and ready for 3.8!

https://bugs.python.org/issue32912

:EDIT:

I don't know how to add a NEWS entry, but I have to go for now. I'll look into that tomorrow.

### CODE REVIEW

File: Python/ast.c [hunk 1]
```diff
@@ -4158,11 +4158,11 @@ warn_invalid_escape_sequence(struct compiling *c, const node *n,
     if (msg == NULL) {
         return -1;
     }
-    if (PyErr_WarnExplicitObject(PyExc_DeprecationWarning, msg,
+    if (PyErr_WarnExplicitObject(PyExc_SyntaxWarning, msg,
                                    c->c_filename, LINENO(n),
                                    NULL, NULL) < 0)
     {
-        if (PyErr_ExceptionMatches(PyExc_DeprecationWarning)) {
+        if (PyErr_ExceptionMatches(PyExc_SyntaxWarning)) {
```
Comment (serhiy-storchaka) [inline Python/ast.c:4165]: See the comment below.

File: Objects/unicodeobject.c [hunk 2]
```diff
@@ -6109,7 +6109,7 @@ PyUnicode_DecodeUnicodeEscape(const char *s,
     if (result == NULL)
         return NULL;
     if (first_invalid_escape != NULL) {
-        if (PyErr_WarnFormat(PyExc_DeprecationWarning, 1,
+        if (PyErr_WarnFormat(PyExc_SyntaxWarning, 1,
```
Comment (serhiy-storchaka) [inline Objects/unicodeobject.c:6112]: This is not related to syntax.

File: Objects/bytesobject.c [hunk 3]
```diff
@@ -1255,7 +1255,7 @@ PyObject *PyBytes_DecodeEscape(const char *s,
     if (result == NULL)
         return NULL;
     if (first_invalid_escape != NULL) {
-        if (PyErr_WarnFormat(PyExc_DeprecationWarning, 1,
+        if (PyErr_WarnFormat(PyExc_SyntaxWarning, 1,
```
Comment (serhiy-storchaka) [inline Objects/bytesobject.c:1258]: This is not related to syntax.

Comment (Vgr255) [reply Objects/bytesobject.c:1258]: This warning was added back in 3.6 when we started deprecating the invalid escapes. In 3.8, we're upgrading these to SyntaxWarning so that they're no longer silent by default, and it will eventually be a SyntaxError. The end goal here is to raise a SyntaxError for this use, but this will have to wait for two more releases.

File: Objects/unicodeobject.c (hunk 2 above)
Comment (Vgr255) [reply Objects/unicodeobject.c:6112]: This warning was added back in 3.6 when we started deprecating the invalid escapes. In 3.8, we're upgrading these to SyntaxWarning so that they're no longer silent by default, and it will eventually be a SyntaxError. The end goal here is to raise a SyntaxError for this use, but this will have to wait for two more releases.

File: Python/ast.c (hunk 1 above)
Comment (Vgr255) [reply Python/ast.c:4165]: Yes, the end goal here is to get a SyntaxError, however this is not yet possible as we have to wait for two releases before this can happen, so we can't yet convert it to a SyntaxError. I probably should update that comment, though (I'll do that tomorrow when I get home).

File: Objects/bytesobject.c (hunk 3 above)
Comment (serhiy-storchaka) [reply Objects/bytesobject.c:1258]: No, this warning is emitted by `codecs.escape_decode()` (which is used in pickle).
```
>>> codecs.escape_decode(b'\\z')
__main__:1: DeprecationWarning: invalid escape sequence '\z'
(b'\\z', 2)
```

File: Objects/unicodeobject.c (hunk 2 above)
Comment (serhiy-storchaka) [reply Objects/unicodeobject.c:6112]: This warning is emitted by the "unicode-escape" codec.
```
>>> b'\\z'.decode('unicode-escape')
__main__:1: DeprecationWarning: invalid escape sequence '\z'
'\\z'
```

File: Objects/bytesobject.c (hunk 3 above)
Comment (Vgr255) [reply Objects/bytesobject.c:1258]: Yes, and this warning [was added][0] as part of bpo-27364 to keep consistency with the rest of the language. It only makes sense to upgrade this warning along with everything else, or else the codecs module falls out of sync with the rest of the language.

[0]: https://bugs.python.org/file44475/deprecate_invalid_escapes_both_5.patch

File: Objects/unicodeobject.c (hunk 2 above)
Comment (Vgr255) [reply Objects/unicodeobject.c:6112]: Yes, and this warning [was added][0] as part of bpo-27364 to keep consistency with the rest of the language. It only makes sense to upgrade this warning along with everything else, or else this codec falls out of sync with the rest of the language.

[0]: https://bugs.python.org/file44475/deprecate_invalid_escapes_both_5.patch

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]
The codecs module shouldn't raise SyntaxWarning or SyntaxError. It should raise DeprecationWarning, which later will be converted to UnicodeDecodeError.

**Comment by Vgr255** [comment]
I haven't had the time to work on this, sorry! I'll get back to it hopefully sometime soon.

**Comment by tirkarthi** [comment]
I think this was done with https://github.com/python/cpython/pull/9652/ and the bpo issue is also marked as resolved.
